pinot-controller/src/main/java/com/linkedin/pinot/controller/validation/ValidationManager.java

```diff
@@ -258,53 +249,44 @@ void validateLLCSegments(final String realtimeTableName, TableConfig tableConfig
   }
 
   // For offline segment pushes, validate that there are no missing segments, and update metrics
-  private void validateOfflineSegmentPush(ZkHelixPropertyStore<ZNRecord> propertyStore, String tableName,
-      List<SegmentMetadata> segmentMetadataList) {
-    List<OfflineSegmentZKMetadata> offlineSegmentZKMetadatas = ZKMetadataProvider
-        .getOfflineSegmentZKMetadataListForTable(propertyStore, tableName);
-    for (OfflineSegmentZKMetadata offlineSegmentZKMetadata : offlineSegmentZKMetadatas) {
-      SegmentMetadata segmentMetadata = new SegmentMetadataImpl(offlineSegmentZKMetadata);
-      segmentMetadataList.add(segmentMetadata);
-    }
-
-    // Calculate missing segments only for offline tables
-    int missingSegmentCount = 0;
+  private void validateOfflineSegmentPush(ZkHelixPropertyStore<ZNRecord> propertyStore, String offlineTableName) {
+    List<OfflineSegmentZKMetadata> offlineSegmentZKMetadataList =
+        ZKMetadataProvider.getOfflineSegmentZKMetadataListForTable(propertyStore, offlineTableName);
 
     // Compute the missing segments if there are at least two
-    if (2 < segmentMetadataList.size()) {
-      List<Interval> segmentIntervals = new ArrayList<Interval>();
-      for (SegmentMetadata segmentMetadata : segmentMetadataList) {
-        Interval timeInterval = segmentMetadata.getTimeInterval();
-        if (timeInterval != null && TimeUtils.timeValueInValidRange(timeInterval.getStartMillis()) && TimeUtils
-                .timeValueInValidRange(timeInterval.getEndMillis())) {
+    int numSegments = offlineSegmentZKMetadataList.size();
+    if (numSegments >= 2) {
+      List<Interval> segmentIntervals = new ArrayList<>(numSegments);
+      Duration timeGranularity = null;
+      for (OfflineSegmentZKMetadata offlineSegmentZKMetadata : offlineSegmentZKMetadataList) {
+        Interval timeInterval = offlineSegmentZKMetadata.getTimeInterval();
+        if (timeInterval != null && TimeUtils.timeValueInValidRange(timeInterval.getStartMillis())
+            && TimeUtils.timeValueInValidRange(timeInterval.getEndMillis())) {
           segmentIntervals.add(timeInterval);
+          timeGranularity = offlineSegmentZKMetadata.getTimeGranularity();
         }
       }
-
-      List<Interval> missingIntervals = computeMissingIntervals(segmentIntervals, segmentMetadataList.get(0).getTimeGranularity());
-      missingSegmentCount = missingIntervals.size();
-
+      List<Interval> missingIntervals = computeMissingIntervals(segmentIntervals, timeGranularity);
       for (Interval missingInterval : missingIntervals) {
-        LOGGER.warn("Missing data in table {} for time interval {}", tableName, missingInterval);
+        LOGGER.warn("Missing data in table {} for time interval {}", offlineTableName, missingInterval);
       }
+      // Update the gauge that contains the number of missing segments
+      _validationMetrics.updateMissingSegmentCountGauge(offlineTableName, missingIntervals.size());
```

Comment:
This needs to be outside the loop, set to 0 if there are no missing segments, right? Otherwise, the gauge for the table, once set, can never be reset to 0

Reply:
My logic here is that, if the number of segments is not enough to support this check, we should not update the count. What do you think?

Reply:
Discussed offline, will move this outside the loop